extensions/message-ids.md

```diff
@@ -0,0 +1,112 @@
+
+The tag value MUST be treated as an opaque identifier.
+
+## Server implementation considerations
+
+This section is non-normative.
+
+In order to guarantee uniqueness, message IDs can't be implemented as simple numeric counters that risk clashing with other servers on the network, or being reset if the server restarts.
+
+Some examples of appropriate IDs are:
+
+* An [RFC 4122 UUID](https://tools.ietf.org/html/rfc4122)
+* An ID generated by a shared external provider, used by all servers on the network
+* A numeric counter, prefixed with a server ID, with the latest value saved externally to the IRCd on every issue, to allow for power failures.
+* A numeric counter, prefixed with a server ID as well as the server startup timestamp at a suitable precision.
+
+When using timestamps, make sure they're correctly synchronised using NTP or similar.
+
+## Client implementation considerations
+
+This section is non-normative.
+
+Message IDs have no guarantee of being universally unique across different IRC networks, nor will they necessarily share the same format. There is also no requirement that numeric IDs increase monotonically. Don't attempt to correlate them beyond their scope and don't use them for message ordering.
+
+## Examples
+
+This section is non-normative, message IDs are not required to be UUIDs or have any specific format. Additional tags used in these examples may or may not have a specified meaning elsewhere.
+
+A channel `PRIVMSG` sent by the server:
+
+    S: @draft/msgid=63E1033A-051D-4B41-B1AB-1FA3CF4B243E :nick!user@host PRIVMSG #channel :Hello!
+
+A private `PRIVMSG` sent by the server:
+
+    S: @draft/msgid=server1-1480339715754191-21 :nick!user@host PRIVMSG me :Hello!
+
+A channel `NOTICE` sent by the server:
+
+    S: @draft/msgid=server2-3 :nick!user@host NOTICE #channel :Hello!
+
+A private `NOTICE` sent by the server:
+
+    S: @draft/msgid=ticketid-5 :nick!user@host NOTICE me :Hello!
+
+A channel `PRIVMSG` sent by the server, and a possible client response. The `+example/reply` tag is a non-standard example:
+
+    S: @draft/msgid=062F321D-425E-4951-A3E8-D644513E2ECD :nick!user@host PRIVMSG #channel :Hello!
+    C: @+example/reply=062F321D-425E-4951-A3E8-D644513E2ECD :nick2!user2@host2 PRIVMSG #channel :Hello to you!
+
+Two channel `PRIVMSG` messages sent by the server, with possible non-standard example annotations to indicate split message concatenation:
+
+    S: @draft/msgid=8FD37885-C4B0-4977-A698-76786BFDDEB2;example/split :nick!user@host PRIVMSG #channel :Hello
+    S: @draft/msgid=DACE5EEB-9BCC-4597-85F4-AB13562992E8;example/concat=8FD37885-C4B0-4977-A698-76786BFDDEB2 :nick!user@host PRIVMSG #channel : World
+
+A client negotiating the `draft/message-tags` capability to enable and disable messages tagged with IDs.
+
+    S: :nick!user@host PRIVMSG #channel :Hello
+    C: CAP REQ draft/message-tags
+    S: :irc.example.com CAP me ACK :draft/message-tags
+    C: CAP END
```

Comment:
Yeah, but our logic is just send it after ACK every time (or start SASL). Extraneous in the examples though so I'll drop it.